### ISSUE TITLE
Support independent Cumulocity MQTT and HTTP endpoint settings (url and port)

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#1931]
**Is your feature request related to a problem? Please describe.**

Some custom Cumulocity setups have two different urls for the MQTT and HTTPS endpoints. Currently thin-edge.io assumes that the two endpoints are on the same url, but on different ports (e.g. MQTT on `8883` and HTTPS on port `443`), which means that thin-edge.io does not work out of the box.

**Describe the solution you'd like**

Support independent settings for Cumulocity endpoints for both MQTT and HTTP(s) endpoints.

Instead of just one `c8y.url`, the following configuration settings would be supported:

```sh
tedge config set c8y.http my.custom.domain.com:443
tedge config set c8y.mqtt device.custom.domain.com:8883
```

* Setting the port as port of the endpoint is optional, however if left blank then the following defaults should be used:
    * `c8y.http`, default to port `443`
    * `c8y.mqtt`, default to port `8883`

**Describe alternatives you've considered**

Alternatively, `c8y.url` could be kept instead of , and use the `c8y.mqtt` only if defined (for better backwards compatibility). This might be the preferred solution.

**Additional context**

Some device owners restrict the number of outgoing ports, so that `443` is the only outgoing port allowed. In this scenario both the http and mqtt traffic would have to be configured to different endpoints (e.g. using a custom nginx setup on the server side)

**Comment by reubenmiller** [comment]
This is partially related to #1834

**Comment by Bravo555** [comment]
We've settled on the following configuration scheme with @reubenmiller:

- `c8y.url`: remains and acts as a default hostname for both HTTP and MQTT endpoints. Cannot contain a port because the ports are implied (443 for HTTPS, 8883 for MQTT)
- **(NEW)** `c8y.http`: Overrides `c8y.url` in places when we connect to HTTP server. Can contain and use a different port than default.
- - **(NEW)** `c8y.mqtt`: Overrides `c8y.url` in places when we connect to MQTT broker. Can contain and use a different port than default.

**Comment by didier-wenzek** [comment]
> We've settled on the following configuration scheme with @reubenmiller:
> 
>     * `c8y.url`: remains and acts as a default hostname for both HTTP and MQTT endpoints. Cannot contain a port because the ports are implied (443 for HTTPS, 8883 for MQTT)
> 
>     * **(NEW)** `c8y.http`: Overrides `c8y.url` in places when we connect to HTTP server. Can contain and use a different port than default.
> 
>     * **(NEW)** `c8y.mqtt`: Overrides `c8y.url` in places when we connect to MQTT broker. Can contain and use a different port than default.


That's perfect.

I would just add a precision. In the code itself, one must no more directly use `c8y.url` but be explicit about whether to connect to c8y over HTTPS or MQTT. The use of the value set with `c8y.url` as the default must be handled by `tedge_config`.

@Bravo555 and @jarhodes314 : be sure to make this working with the [new tedge config](https://github.com/thin-edge/thin-edge.io/pull/1936).

**Comment by jarhodes314** [comment]
> > We've settled on the following configuration scheme with @reubenmiller:
> > 
> >     * `c8y.url`: remains and acts as a default hostname for both HTTP and MQTT endpoints. Cannot contain a port because the ports are implied (443 for HTTPS, 8883 for MQTT)
> > 
> >     * **(NEW)** `c8y.http`: Overrides `c8y.url` in places when we connect to HTTP server. Can contain and use a different port than default.
> > 
> >     * **(NEW)** `c8y.mqtt`: Overrides `c8y.url` in places when we connect to MQTT broker. Can contain and use a different port than default.
> 
> 
> That's perfect.
> 
> I would just add a precision. In the code itself, one must no more directly use `c8y.url` but be explicit about whether to connect to c8y over HTTPS or MQTT. The use of the value set with `c8y.url` as the default must be handled by `tedge_config`.
> 
> @Bravo555 and @jarhodes314 : be sure to make this working with the [new tedge config](https://github.com/thin-edge/thin-edge.io/pull/1936).

So to clarify this, you would like c8y.url to exist like it does now, but not be exposed to crates to read (i.e. the field will be private), and then c8y.mqtt.url and c8y.http.url can either be set as configurations in their own right, or they use c8y.url as a fallback.

**Comment by didier-wenzek** [comment]
> So to clarify this, you would like c8y.url to exist like it does now, but not be exposed to crates to read (i.e. the field will be private), and then c8y.mqtt.url and c8y.http.url can either be set as configurations in their own right, or they use c8y.url as a fallback?

That's correct. Except that I'm okay if `c8y.url` is not private (I don't want to have something too complex in the tedge_config crate to handle this exception), what matters is that the default logic (use url + default port, if mqtt is no set) is hidden behind c8y.mqtt.

**Comment by Bravo555** [comment]
> So to clarify this, you would like c8y.url to exist like it does now, but not be exposed to crates to read (i.e. the field will be private), and then c8y.mqtt.url and c8y.http.url can either be set as configurations in their own right, or they use c8y.url as a fallback?

That's what I also wanted to do. Under the current tedge_config, I changed `C8yUrlSetting::query` to always error saying this accessor was deprecated, and removed it from `config_key.rs` in `tedge`, but that breaks `config set` and `config list` commands because they implicitly depend on the `query` method working. Being able to remove a getter intended for client code and leaving `set` and `list` options as they are would be ideal.

For now, all the places where `C8yUrlSetting` was used were replaced or removed anyway, so I left the `query` method as it is, but we need some way to prohibit use of deprecated settings/getters in new code.
